utils/spdx/src/main/kotlin/parser/SpdxExpressionLexer.kt

```diff
@@ -0,0 +1,103 @@
+
+/**
+ * A lexer for SPDX expressions. It consumes a sequence of characters and produces a sequence of [Token]s. For details
+ * on the grammar see [SpdxExpressionParser].
+ */
+class SpdxExpressionLexer(input: Sequence<Char>) {
+    constructor(input: String) : this(input.asSequence())
+
+    companion object {
+        /** The uppercase characters allowed by the SPDX specification. */
+        private val UPPERCASE = 'A'..'Z'
+        /** The lowercase characters allowed by the SPDX specification. */
+        private val LOWERCASE = 'a'..'z'
+        /** The digits allowed by the SPDX specification. */
+        private val DIGITS = '0'..'9'
+
+        /** Return true if the character is an uppercase character allowed by the SPDX specification. */
+        private fun Char.isUpper() = this in UPPERCASE
+        /** Return true if the character is a lowercase character allowed by the SPDX specification. */
+        private fun Char.isLower() = this in LOWERCASE
+        /** Return true if the character is a digit allowed by the SPDX specification. */
+        private fun Char.isDigit() = this in DIGITS
+        /** Return true if the character is a valid identifier character allowed by the SPDX specification. */
+        private fun Char.isIdentifier() = isUpper() || isLower() || isDigit() || this == '.' || this == '-'
+    }
+
+    private val iterator = input.iterator()
+    private var next = iterator.nextOrNull()
+    private var position = 0
+
+    fun tokens(): Sequence<Token> = generateSequence { nextToken() }
+
+    private fun nextToken(): Token? {
+        var cur = consumeChar()
+
+        while (cur != null) {
+            if (cur == ' ') {
+                cur = consumeChar()
+                continue
+            }
+
+            if (cur == '(') return Token.OPEN(position)
+            if (cur == ')') return Token.CLOSE(position)
+            if (cur == '+') return Token.PLUS(position)
+            if (cur == ':') return Token.COLON(position)
+
+            if (cur.isIdentifier()) {
+                val start = position
+                val value = buildString {
+                    append(cur)
+
+                    while (next?.isIdentifier() == true) {
+                        cur = consumeChar()
+                        append(cur)
+                    }
+                }
+
+                return when (value.uppercase()) {
+                    "AND" -> Token.AND(start)
+                    "OR" -> Token.OR(start)
+                    "WITH" -> Token.WITH(start)
+                    else -> when {
+                        value.startsWith("DocumentRef-") -> Token.DOCUMENTREF(start, position, value)
+                        value.startsWith("LicenseRef-") -> Token.LICENSEREF(start, position, value)
+                        else -> Token.IDENTIFIER(start, position, value)
+                    }
+                }
+            }
+
+            cur?.let { throw SpdxExpressionLexerException(it, position) }
```

Comment:
Nit: Maybe `also` is more fitting as the return value of the block does not matter?

Reply:
Or maybe just use pain `if` to align with line 90 of `SpdxExpressionParser.kt`.

Reply:
Somehow the type resolved is confused here, for the line 

```kotlin
if (cur != null) throw SpdxExpressionLexerException(cur, position)
```

it says `Condition 'cur != null' is always true` but also `Smart cast to 'Char' is impossible, because 'cur' is a local variable that is captured by a changing closure`. So I went with the current line. Maybe that's a detail we can figure out after the merge.

Reply:
BTW, I believe this is a kind of bug that will go away with the K2 Kotlin compiler.